pkg/sentry/devices/memdev/kmsg.go

```diff
@@ -0,0 +1,140 @@
+// limitations under the License.
+
+package memdev
+
+import (
+	"bytes"
+	"io"
+	"sync"
+
+	"gvisor.dev/gvisor/pkg/abi/linux"
+	"gvisor.dev/gvisor/pkg/context"
+	"gvisor.dev/gvisor/pkg/safemem"
+	"gvisor.dev/gvisor/pkg/sentry/vfs"
+	"gvisor.dev/gvisor/pkg/syserr"
+	"gvisor.dev/gvisor/pkg/syserror"
+	"gvisor.dev/gvisor/pkg/usermem"
+)
+
+const (
+	BUFFER_ENTRY_MAX = 512
+	BUFFER_MAX       = 1024
+)
+
+// kmsgDevice implements vfs.Device for /dev/kmsg.
+type kmsgDevice struct{}
+
+func (kmsgDevice) Open(ctx context.Context, mnt *vfs.Mount, vfsd *vfs.Dentry, opts vfs.OpenOptions) (*vfs.FileDescription, error) {
+	fd := &kmsgFD{newmsg: make(chan int)}
+	if err := fd.vfsfd.Init(fd, opts.Flags, mnt, vfsd, &vfs.FileDescriptionOptions{
+		UseDentryMetadata: true,
+	}); err != nil {
+		return nil, err
+	}
+	return &fd.vfsfd, nil
+}
+
+// kmsgFD implements vfs.FileDescriptionImpl for /dev/kmsg.
+type kmsgFD struct {
+	vfsfd vfs.FileDescription
+	vfs.FileDescriptionDefaultImpl
+	vfs.DentryMetadataFileDescriptionImpl
+	vfs.NoLockFD
+
+	// buffer is a circular buffer that store user's write to kmsg.
+	m            sync.Mutex
+	buffer       [BUFFER_ENTRY_MAX]*bytes.Buffer
+	readIndex    int
+	writeIndex   int
+	bufferLength int
+	newmsg       chan int
+}
+
+// Release implements vfs.FileDescriptionImpl.Release.
+func (fd *kmsgFD) Release() {
+	// noop
+}
+
+// Read implements vfs.FileDescriptionImpl.Read.
+func (fd *kmsgFD) Read(ctx context.Context, dst usermem.IOSequence, opts vfs.ReadOptions) (int64, error) {
+RETRY:
+	fd.m.Lock()
+
+	if fd.bufferLength > len(fd.buffer) {
+		fd.readIndex = fd.writeIndex
+		fd.bufferLength = len(fd.buffer)
+		fd.m.Unlock()
+		return 0, syserr.ErrBrokenPipe.ToError()
+	}
+	if fd.bufferLength == 0 {
+		fd.m.Unlock()
+		if fd.vfsfd.StatusFlags()&^linux.O_NONBLOCK != 0 {
+			return 0, syserror.ErrWouldBlock
+		}
+		<-fd.newmsg
+		goto RETRY
+	}
+	bytesCopied, err := dst.CopyOutFrom(ctx, safemem.FromIOReader{bytes.NewReader(fd.buffer[fd.readIndex].Bytes())})
+	fd.readIndex++
+	fd.bufferLength--
+	if fd.readIndex == len(fd.buffer) {
+		fd.readIndex = 0
+	}
+	fd.m.Unlock()
+	return bytesCopied, err
+}
+
+// Write implements vfs.FileDescriptionImpl.Write.
+func (fd *kmsgFD) Write(ctx context.Context, src usermem.IOSequence, opts vfs.WriteOptions) (int64, error) {
+	if src.NumBytes() > BUFFER_MAX {
+		return 0, syserr.ErrInvalidArgument.ToError()
+	}
+
+	fd.m.Lock()
+	defer fd.m.Unlock()
+	fd.buffer[fd.writeIndex] = new(bytes.Buffer)
+	bytesCopied, err := src.CopyInTo(ctx, safemem.FromIOWriter{io.Writer(fd.buffer[fd.writeIndex])})
+	fd.writeIndex++
+	fd.bufferLength++
+	if fd.writeIndex == len(fd.buffer) {
+		fd.writeIndex = 0
+	}
+	fd.newmsg <- 1
+	return bytesCopied, err
+}
+
+func (fd *kmsgFD) Seek(ctx context.Context, offset int64, whence int32) (int64, error) {
+	if offset != 0 {
+		return 0, syserr.ErrIllegalSeek.ToError()
+	}
+
+	fd.m.Lock()
+	defer fd.m.Unlock()
+	switch whence {
```

Comment:
The seek behaviour for kmsg is pretty unusual compared to normal files. Could you add comments saying so, and summarizing what each case is supposed to do?